File: 04-17-2024/Tree.java
import java.util.LinkedList;
import java.util.Queue;

public class Tree {
    public static TreeNode root;
    public Tree() {
        root = null;
    }

    public TreeNode arrayToTree(Integer[] input){
        root = createTreeNode(input,0);
        return root;
    }

    private TreeNode createTreeNode(Integer[] input, int index){
        if (index >= input.length) {
            return null;
        }
        TreeNode node = null;
        if (input[index] != null) {
            node = new TreeNode(input[index]);
            node.left = createTreeNode(input, 2 * index + 1); // Left child
            node.right = createTreeNode(input, 2 * index + 2); // Right child
        }
        return node;
    }

    // Function to print the tree using inorder traversal
// Function to print the tree nodes in level-order traversal
    public void printTree(TreeNode rooot) {
        if(rooot == null)
            return;

        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(rooot);

        while (!queue.isEmpty()) {
            TreeNode node = queue.poll();
            System.out.print(node.val);
            if (!queue.isEmpty())
                System.out.print(", ");

            if (node.left != null)
                queue.add(node.left);
            if (node.right != null)
                queue.add(node.right);
        }
    }
}

File: 2025-09-29/Tree.py
import collections
from TreeNode import TreeNode


class Tree:
    def __init__(self, arr=None):
        if arr:
            self.root = self.array_to_tree(arr)
        else:
            self.root = None

    def array_to_tree(self, arr):
        if not arr:
            return None

        # Create the root node from the first element
        root = TreeNode(arr[0])

        # Use a queue to keep track of parent nodes
        queue = collections.deque([root])
        i = 1  # Start from the second element for the children

        while queue and i < len(arr):
            current_node = queue.popleft()

            # Process the left child
            if arr[i] is not None:
                left_child = TreeNode(arr[i])
                current_node.left = left_child
                queue.append(left_child)
            i += 1

            if i >= len(arr):
                break

            # Process the right child
            if arr[i] is not None:
                right_child = TreeNode(arr[i])
                current_node.right = right_child
                queue.append(right_child)
            i += 1

        return root

    def print_tree(self):
        if not self.root:
            print("Tree is empty.")
            return
        self._print_recursive(self.root, "", True)

    def _print_recursive(self, node, prefix, is_left):
        if not node:
            return

        # Print the right subtree first
        if node.right:
            self._print_recursive(node.right, prefix + ("│   " if is_left else "    "), False)

        # Print the current node
        print(prefix + ("└── " if is_left else "┌── ") + str(node.val))

        # Print the left subtree
        if node.left:
            self._print_recursive(node.left, prefix + ("    " if is_left else "│   "), True)
